File: app/src/main/java/com/tanish2k09/sce/SettingsActivity.java
package com.tanish2k09.sce;

import android.content.SharedPreferences;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import androidx.cardview.widget.CardView;
import android.widget.Switch;
import android.widget.Toast;
import com.pes.androidmaterialcolorpickerdialog.ColorPicker;

public class SettingsActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);
        getWindow().setNavigationBarColor(ContextCompat.getColor(this, R.color.black));

        SharedPreferences sp = getSharedPreferences("settings",MODE_PRIVATE);

        Switch autoImportConfig = findViewById(R.id.autoImportConfigSwitch);
        Switch autoUpdateConfig = findViewById(R.id.autoUpdateConfigSwitch);
        Switch useTitles = findViewById(R.id.useTitlesOnCards);
        Switch useBlackBG = findViewById(R.id.useBlackNotDark);
        CardView colorCard = findViewById(R.id.colorCard);

        autoImportConfig.setChecked(sp.getBoolean("autoImportConfig", false));
        autoImportConfig.setOnCheckedChangeListener((buttonView, isChecked) -> {
            sp.edit().putBoolean("autoImportConfig", autoImportConfig.isChecked()).apply();
            Toast.makeText(this, "autoImportConfig: Restart app to apply", Toast.LENGTH_SHORT).show();
        });

        autoUpdateConfig.setChecked(sp.getBoolean("autoUpdateConfig", false));
        autoUpdateConfig.setOnCheckedChangeListener((buttonView, isChecked) -> sp.edit().putBoolean("autoUpdateConfig", autoUpdateConfig.isChecked()).apply());

        useTitles.setChecked(sp.getBoolean("useTitlesOnCards", false));
        useTitles.setOnCheckedChangeListener((buttonView, isChecked) -> sp.edit().putBoolean("useTitlesOnCards", isChecked).apply());

        useBlackBG.setChecked(sp.getBoolean("useBlackNotDark", true));
        setThemeColor(useBlackBG.isChecked());
        useBlackBG.setOnCheckedChangeListener(((buttonView, isChecked) -> {
            sp.edit().putBoolean("useBlackNotDark", isChecked).apply();

            String color = "#121212";

            if (isChecked)
                color = "#000000";

            setThemeColor(color);
        }));

        setColorCard(colorCard, Color.parseColor(sp.getString("accentCol", "#00bfa5")));
        colorCard.setOnClickListener(v -> {
            int preColor = Color.parseColor(sp.getString("accentCol", "#00bfa5"));
            final ColorPicker cp = new ColorPicker(this,Color.red(preColor), Color.green(preColor), Color.blue(preColor));
            cp.enableAutoClose();
            cp.setCallback(color -> {
                sp.edit().putString("accentCol", String.format("#%06X", (0xFFFFFF & color))).apply();
                setColorCard(colorCard, color);
            });
            cp.show();
        });
    }

    private void setColorCard(CardView card, int color) {
        card.setCardBackgroundColor(color);
    }

    private void setThemeColor(String colorHex) {
        ConstraintLayout settingsLayout = findViewById(R.id.settingsLayout);

        ColorDrawable themeColor = new ColorDrawable(Color.parseColor(colorHex));
        settingsLayout.setBackground(themeColor);
        getWindow().setStatusBarColor(Color.parseColor(colorHex));
    }

    private void setThemeColor(boolean isBlackChecked) {
        if (isBlackChecked)
            setThemeColor("#000000");
        else
            setThemeColor("#121212");
    }
}

File: app/src/main/java/com/tanish2k09/sce/SettingsActivity.kt
package com.tanish2k09.sce

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.graphics.Color
import android.os.Bundle
import android.view.View
import android.view.ViewAnimationUtils
import android.view.ViewTreeObserver
import android.view.animation.AccelerateInterpolator
import android.view.animation.DecelerateInterpolator
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.marginTop
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.tanish2k09.sce.data.enums.ETheme
import com.tanish2k09.sce.databinding.ActivitySettingsBinding
import com.tanish2k09.sce.utils.extensions.changeTopMargin
import com.tanish2k09.sce.viewmodels.SharedPrefsVM
import kotlin.math.hypot


class SettingsActivity : AppCompatActivity(){

    private var revealX = 0
    private var revealY = 0
    private lateinit var binding: ActivitySettingsBinding
    private lateinit var settingsVM: SharedPrefsVM

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (savedInstanceState == null)
            overridePendingTransition(R.anim.no_op, R.anim.no_op)

        binding = ActivitySettingsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        settingsVM = ViewModelProvider(this).get(SharedPrefsVM::class.java)
        attachViewModelObservers()

        initUIListeners()

        if (intent.hasExtra("x")) {
            revealX = intent.getIntExtra("x", 0)
        }

        if (intent.hasExtra("y")) {
            revealY = intent.getIntExtra("y", 0)
        }

        if (savedInstanceState != null) {
            return
        }

        binding.settingsLayout.visibility = View.INVISIBLE
        val vto: ViewTreeObserver = binding.settingsLayout.viewTreeObserver

        if (vto.isAlive.not()) {
            return
        }

        vto.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                revealActivity(revealX, revealY)
                binding.settingsLayout.viewTreeObserver.removeOnGlobalLayoutListener(this)
            }
        })
    }

    override fun onResume() {
        super.onResume()
        settingsVM.readSettingsToFields()
    }

    private fun initUIListeners() {
        binding.appTitle.setOnApplyWindowInsetsListener {view, insets ->
            changeTopMargin(view, binding.appTitle.marginTop + insets.systemWindowInsetTop)
            insets
        }

        binding.autoImportConfigSwitch.setOnCheckedChangeListener { _, b ->
            settingsVM.setAutoImport(b)
        }

        binding.autoScriptSwitch.setOnCheckedChangeListener { _, b ->
            settingsVM.setRunScript(b)
        }

        binding.useTitlesOnCards.setOnCheckedChangeListener { _, b ->
            settingsVM.setUseTitles(b)
        }

        binding.useBlackNotDark.setOnCheckedChangeListener { _, b ->
            if (b) {
                settingsVM.setTheme(ETheme.BLACK)
            } else {
                settingsVM.setTheme(ETheme.DARK)
            }
        }
    }

    private fun attachViewModelObservers() {
        settingsVM.autoImport.observe(this, Observer {
            binding.autoImportConfigSwitch.isChecked = it
        })

        settingsVM.runScript.observe(this, Observer {
            binding.autoScriptSwitch.isChecked = it
        })

        settingsVM.useTitles.observe(this, Observer {
            binding.useTitlesOnCards.isChecked = it
        })

        settingsVM.theme.observe(this, Observer {
            binding.useBlackNotDark.isChecked = (it == ETheme.BLACK)
            handleTheme(it)
        })
    }

    private fun handleTheme(newTheme: ETheme) {
        val mainColor = Color.parseColor(newTheme.hex)
        binding.settingsLayout.setBackgroundColor(mainColor)
    }

    private fun revealActivity(x: Int, y: Int) {
        val finalRadius = hypot(
                binding.settingsLayout.width.toDouble(),
                binding.settingsLayout.height.toDouble()).toFloat()

        // create the animator for this view (the start radius is zero)
        val circularReveal = ViewAnimationUtils.createCircularReveal(
                binding.settingsLayout,
                x, y,
                0f, finalRadius
        )

        circularReveal.duration = 500
        circularReveal.interpolator = AccelerateInterpolator()
        // make the view visible and start the animation
        binding.settingsLayout.visibility = View.VISIBLE
        circularReveal.start()
    }

    private fun unRevealActivity() {
        val finalRadius = hypot(
                binding.settingsLayout.width.toDouble(),
                binding.settingsLayout.height.toDouble()).toFloat()
        val circularReveal = ViewAnimationUtils.createCircularReveal(
                binding.settingsLayout, revealX, revealY, finalRadius, 0f)
        circularReveal.duration = 500
        circularReveal.interpolator = DecelerateInterpolator()
        circularReveal.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                binding.settingsLayout.visibility = View.INVISIBLE
                finish()
            }
        })
        circularReveal.start()
    }

    override fun onBackPressed() {
        unRevealActivity()
    }
}
